File: IrrgartenP2/IrrgartenJava/src/irrgartenjava/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgartenjava;

/**
 *
 * @author moham
 */// Labyrinth.java
import java.util.ArrayList;

public class Labyrinth {
    private static final char BLOCK_CHAR_OBSTACLE = 'X';
    private static final char BLOCK_CHAR_EMPTY = '-';
    private static final char BLOCK_CHAR_MONSTER = 'M';
    private static final char BLOCK_CHAR_EXIT = 'E';
    private static final char BLOCK_CHAR_COMBAT = 'C';

    private static final int ROW = 0;
    private static final int COL = 1;

    private int nRows, nCols;
    private int exitRow, exitCol; // Exit coordinates
    private Monster[][] monsters;
    private Player[][] players;
    private char[][] cells;

    public Labyrinth(int nRows, int nCols) {
        this.nRows = nRows;
        this.nCols = nCols;
        this.monsters = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.cells = new char[nRows][nCols];

        // Initialize all cells to empty
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                cells[i][j] = BLOCK_CHAR_EMPTY;
            }
        }

        // Set initial exit position to an invalid location
        this.exitRow = -1;
        this.exitCol = -1;
    }

    public void setExit(int row, int col){
        this.exitRow = row;
        this.exitCol = col;
    }
    
    public boolean haveAWinner() {
        // Check if any player is at the exit position
        return players[exitRow][exitCol] != null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                sb.append(cells[i][j]).append(' ');
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    public void addMonster(int row, int col, Monster monster) {
        if (posOK(row, col) && cells[row][col] == BLOCK_CHAR_EMPTY) {
            monsters[row][col] = monster;
            monster.setPos(row, col);
            cells[row][col] = BLOCK_CHAR_MONSTER;
        }
    }

    public boolean posOK(int row, int col) {
        return row >= 0 && row < nRows && col >= 0 && col < nCols;
    }

    public boolean emptyPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_EMPTY;
    }

    public boolean monsterPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_MONSTER;
    }

    public boolean exitPos(int row, int col) {
        return row == exitRow && col == exitCol;
    }

    public boolean combatPos(int row, int col) {
        return cells[row][col] == BLOCK_CHAR_COMBAT;
    }

    public boolean canStepOn(int row, int col) {
        return posOK(row, col) && (emptyPos(row, col) || monsterPos(row, col) || exitPos(row, col));
    }

    public void updateOldPos(int row, int col) {
        if (posOK(row, col)) {
            if (combatPos(row, col)) {
                cells[row][col] = BLOCK_CHAR_MONSTER;
            } else {
                cells[row][col] = BLOCK_CHAR_EMPTY;
            }
        }
    }

    public int[] dir2Pos(int row, int col, Directions direction) {
        int[] newPos = {row, col};
        switch (direction) {
            case UP: newPos[ROW]--; break;
            case DOWN: newPos[ROW]++; break;
            case LEFT: newPos[COL]--; break;
            case RIGHT: newPos[COL]++; break;
        }
        return newPos;
    }

    public int[] randomEmptyPos() {
        int[] position;
        do {
            int row = Dice.randomPos(nRows);
            int col = Dice.randomPos(nCols);
            position = new int[]{row, col};
        } while (!emptyPos(position[ROW], position[COL]));
        return position;
    }

    public void spreadPlayers(ArrayList<Player> players) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public Monster putPlayer(Directions direction, Player player) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public void addBlock(Orientation orientation, int startRow, int startCol, int length) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }

    public ArrayList<Directions> validMoves(int row, int col) {
        throw new UnsupportedOperationException("Method not implemented yet.");
    }
}

File: IrrgartenP2/irrgartenRuby/labyrinth.rb
# labyrinth.rb
class Labyrinth
  BLOCK_CHAR_OBSTACLE = 'X'
  BLOCK_CHAR_EMPTY = '-'
  BLOCK_CHAR_MONSTER = 'M'
  BLOCK_CHAR_EXIT = 'E'
  BLOCK_CHAR_COMBAT = 'C'

  ROW = 0
  COL = 1

  attr_reader :n_rows, :n_cols, :exit_row, :exit_col

  def initialize(n_rows, n_cols)
    @n_rows = n_rows
    @n_cols = n_cols
    @monsters = Array.new(n_rows) { Array.new(n_cols) }
    @players = Array.new(n_rows) { Array.new(n_cols) }
    @cells = Array.new(n_rows) { Array.new(n_cols, BLOCK_CHAR_EMPTY) }
    @exit_row = -1
    @exit_col = -1
  end

  def set_exit(row, col)
    @exit_row = row
    @exit_col = col
  end

  def have_a_winner?
    !@players[@exit_row][@exit_col].nil?
  end

  def to_s
    @cells.map { |row| row.join(' ') }.join("\n")
  end

  def add_monster(row, col, monster)
    if pos_ok?(row, col) && @cells[row][col] == BLOCK_CHAR_EMPTY
      @monsters[row][col] = monster
      monster.set_pos(row, col)
      @cells[row][col] = BLOCK_CHAR_MONSTER
    end
  end

  def pos_ok?(row, col)
    row.between?(0, @n_rows - 1) && col.between?(0, @n_cols - 1)
  end

  def empty_pos?(row, col)
    @cells[row][col] == BLOCK_CHAR_EMPTY
  end

  def monster_pos?(row, col)
    @cells[row][col] == BLOCK_CHAR_MONSTER
  end

  def exit_pos?(row, col)
    row == @exit_row && col == @exit_col
  end

  def combat_pos?(row, col)
    @cells[row][col] == BLOCK_CHAR_COMBAT
  end

  def can_step_on?(row, col)
    pos_ok?(row, col) && (empty_pos?(row, col) || monster_pos?(row, col) || exit_pos?(row, col))
  end

  def update_old_pos(row, col)
    if pos_ok?(row, col)
      @cells[row][col] = combat_pos?(row, col) ? BLOCK_CHAR_MONSTER : BLOCK_CHAR_EMPTY
    end
  end

  def dir_to_pos(row, col, direction)
    case direction
    when :up then [row - 1, col]
    when :down then [row + 1, col]
    when :left then [row, col - 1]
    when :right then [row, col + 1]
    else [row, col]
    end
  end

  def random_empty_pos
    loop do
      row = Dice.random_pos(@n_rows)
      col = Dice.random_pos(@n_cols)
      return [row, col] if empty_pos?(row, col)
    end
  end

  def spread_players(players)
    raise NotImplementedError, "Method not implemented yet."
  end

  def put_player(direction, player)
    raise NotImplementedError, "Method not implemented yet."
  end

  def add_block(orientation, start_row, start_col, length)
    raise NotImplementedError, "Method not implemented yet."
  end

  def valid_moves(row, col)
    raise NotImplementedError, "Method not implemented yet."
  end
end
